source/io/Bootstrap.cpp

```diff
@@ -20,35 +20,74 @@ namespace Aws
+            // Lives until the bootstrap's shutdown-complete callback fires.
+            struct ClientBootstrapCallbackData
+            {
+                std::promise<void> ShutdownPromise;
+                OnClientBootstrapShutdownComplete ShutdownCallback;
+
+                static void OnShutdownComplete(void *userData)
+                {
+                    auto callbackData = static_cast<ClientBootstrapCallbackData *>(userData);
+
+                    callbackData->ShutdownPromise.set_value();
+                    callbackData->ShutdownCallback();
+
+                    delete callbackData;
+                }
+            };
+
             ClientBootstrap::ClientBootstrap(
                 EventLoopGroup &elGroup,
                 HostResolver &resolver,
                 Allocator *allocator) noexcept
-                : m_lastError(AWS_ERROR_SUCCESS)
+                : m_bootstrap(nullptr), m_lastError(AWS_ERROR_SUCCESS),
+                  m_callbackData(new ClientBootstrapCallbackData()), m_enableBlockingShutdown(false)
             {
+                m_shutdownFuture = m_callbackData->ShutdownPromise.get_future();
+
                 aws_client_bootstrap_options options;
                 options.event_loop_group = elGroup.GetUnderlyingHandle();
                 options.host_resolution_config = resolver.GetConfig();
                 options.host_resolver = resolver.GetUnderlyingHandle();
-                options.on_shutdown_complete = NULL;
-                options.user_data = NULL;
+                options.on_shutdown_complete = ClientBootstrapCallbackData::OnShutdownComplete;
+                options.user_data = m_callbackData.get();
                 m_bootstrap = aws_client_bootstrap_new(allocator, &options);
+                if (!m_bootstrap)
+                {
+                    m_lastError = aws_last_error();
+                }
             }
 
             ClientBootstrap::~ClientBootstrap()
             {
                 if (m_bootstrap)
                 {
+                    // Release m_callbackData, it destroys itself when shutdown completes.
+                    m_callbackData.release();
+
                     aws_client_bootstrap_release(m_bootstrap);
-                    m_bootstrap = nullptr;
-                    m_lastError = AWS_ERROR_UNKNOWN;
+                    if (m_enableBlockingShutdown)
+                    {
+                        // If your program is stuck here, stop using EnableBlockingShutdown()
+                        m_shutdownFuture.wait();
+                    }
                 }
             }
 
             ClientBootstrap::operator bool() const noexcept { return m_lastError == AWS_ERROR_SUCCESS; }
 
             int ClientBootstrap::LastError() const noexcept { return m_lastError; }
 
+            void ClientBootstrap::SetShutdownCompleteCallback(OnClientBootstrapShutdownComplete callback)
+            {
+                m_callbackData->ShutdownCallback = std::move(callback);
```

Comment:
I trust your judgement on this one, but shouldn't callback be an r-value?

Reply:
This felt right to me. We're storing a copy, so we take this by copy (then we move it into our member).

Couldn't find a satisfying answer on Stackoverflow on the **one true way**

Ran both variants through godbolt and ended up with the same generated code so 🤷‍♀️